Sources/SignalRClient/MessageBuffer.swift

```diff
@@ -0,0 +1,139 @@
+
+import Foundation
+
+actor MessageBuffer {
+    private var maxBufferSize: Int
+    private var messages: [BufferedItem] = []
+    private var bufferedByteCount: Int = 0
+    private var totalMessageCount: Int = 0
+    private var lastSendSequenceId: Int = 0
+    private var lastSendIdx = 0
+    private var dequeueContinuation: CheckedContinuation<Bool, Never>?
+    private var closed: Bool = false
+
+    init(bufferSize: Int) {
+        self.maxBufferSize = bufferSize
+    }
+
+    public func enqueue(content: StringOrData) async throws -> Void {
+        if closed {
+            throw SignalRError.invalidOperation("Message buffer has closed")
+        }
+
+        var size: Int
+        switch content {
+        case .string(let str):
+            size = str.lengthOfBytes(using: .utf8)
+        case .data(let data):
+            size = data.count
+        }
+
+        bufferedByteCount = bufferedByteCount + size
+        totalMessageCount = totalMessageCount + 1
+
+        return await withCheckedContinuation{ continuation in
+            if (bufferedByteCount > maxBufferSize) {
+                // If buffer is full, we're tring to backpressure the sending
+                // id start from 1
+                messages.append(BufferedItem(content: content, size: size, id: totalMessageCount, continuation: continuation))
+            } else {            
+                messages.append(BufferedItem(content: content, size: size, id: totalMessageCount, continuation: nil))
+                continuation.resume()
+            }
+
+            if let continuation = dequeueContinuation {
+                continuation.resume(returning: true)
+            }
+        }
+    }
+
+    public func ack(sequenceId: Int) throws -> Bool {
+        // It might be wrong ack or the ack of previous connection
+        if (sequenceId <= 0 || sequenceId > lastSendSequenceId) {
+            return false
+        }
+
+        let pfx = messages.prefix {$0.id <= sequenceId}
+        let itemsToProcess = Array(pfx)
+        // make sure remove before any async operation for concurrency issue
+        messages = Array(messages.dropFirst(pfx.count))
+        // sending idx will change because we changes the array
+        lastSendIdx = lastSendIdx - pfx.count
+
+        guard lastSendIdx >= 0 else {
+            throw SignalRError.invalidOperation("Index of the ack < 0, fatal error")
+        }
+
+        for item in itemsToProcess {
+            bufferedByteCount = bufferedByteCount - item.size
+            if let ctu = item.continuation {
+                ctu.resume()
+            }
+        }
+        return true
+    }
+
+    public func WaitToDequeue() async throws -> Bool {
+        let lastEnqueuedIdx = messages.count - 1
+        if (lastSendIdx <= lastEnqueuedIdx) {
+            return true
+        }
+        return await withCheckedContinuation { continuation in
+            dequeueContinuation = continuation
```

Comment:
This method requires only one can `WaitToDequeue`. Otherwize the `dequeueContinuation` will be leaked

Reply:
Improved